File: 2/MinMaxStack.java
import java.util.*;

class MinMaxStack {
	Stack<Integer> stack = new Stack<Integer>();
	Stack<HashMap<String, Integer>> minMaxStack = new Stack<>();

	public int getMax() {
		return minMaxStack.peek().get("Maximum");
	}

	public int getMin() {
		return minMaxStack.peek().get("Minimum");
	}
	
	public void push(int data) {
		if(minMaxStack.isEmpty()) {
			HashMap<String, Integer> details = new HashMap<String, Integer>();
			details.put("Maximum", data);
			details.put("Minimum", data);
			minMaxStack.push(details);
		} else {
				HashMap<String, Integer> details = new HashMap<String, Integer>();
				details.put("Maximum", Math.max(getMax(), data));
				details.put("Minimum", Math.min(getMin(), data));
				minMaxStack.push(details);
		}

		stack.push(data);
	}
	
	public void pop() {
		stack.pop();
		minMaxStack.pop();
	}
	
	public int peek() {
		return stack.peek();
	}

	public static void main(String args[]) {
		MinMaxStack stack = new MinMaxStack();
		stack.push(5);
		stack.push(7);
		stack.push(2);

		System.out.println(stack.peek());
		System.out.println(stack.getMax());
		System.out.println(stack.getMin());
		
	}
}

File: 2/js/src/MinMaxStack.ts
interface Frame {
	cur: number;
	max: number;
	min: number;
}

class MinMaxStack {
	stack: Frame[]

	constructor() {
		this.stack = [];
	}

	peek(): Frame {
		return this.stack[this.stack.length-1];
	}

	pop(): Frame | undefined {
		return this.stack.pop();
	}

	push(data: number) {
		if(this.peek() == undefined) {
			this.stack.push({cur: data, max: data, min: data});
		} else {
			this.stack.push({cur: data, max: Math.max(this.getMax(), data), min: Math.min(this.getMin(), data)});
		}
	}

	getMax() {
		const frame = this.peek();
		if(frame) return frame.max;
		return -1;
	}

	getMin() {
		const frame = this.peek();
		if(frame) return frame.min;
		return -1;
	}
}
